sycl/include/sycl/queue.hpp

```diff
@@ -54,6 +54,14 @@
 #define __SYCL_USE_FALLBACK_ASSERT 0
 #endif
 
+// The pi_native_handle2 type has been added as a temporary measure so that the
+// existing pi_native_handle can co-exist with it. At the next ABI redefinition
+// the "2" version will be removed and only pi_native_handle will remain.
+typedef struct _ze_command_queue_handle_t *ze_command_queue_handle_t;
+typedef struct _ze_command_list_handle_t *ze_command_list_handle_t;
+using pi_native_handle2 =
+    std::variant<ze_command_queue_handle_t, ze_command_list_handle_t>;
```

Comment:
Can't these be just 
```suggestion
using pi_native_handle2 = std::variant<void*, void*>;
```

Reply:
The std::variant explicitly prohibits the void* type.
The value returned by get_native must be queried for which type it is: command queue or command list. For that to work, we need to expose the underlying types. We are trying to use modern C++ features here, rather than a union, a boolean flag, and casting.

Reply:
It seems `std::variant` prohibits `void` type only, the `void *` is allowed.
I believe this is undesirable to mention Level Zero handles in API agnostic code.
What will happen with this function when the ABI break is allowed and how it plays with OpenCL and CUDA interoperability?

Reply:
also you can try `std::variant<uintptr_t, uintptr_t>`

Reply:
Yes, essentially <pi_native_handle, pi_native_handle>. Let me try that.
The user interface to make_queue / get_native will be more awkward with this type-agnostic way. There will have to be a convention that index 0 of the variant is command queue and index 1 is command list and the user will have to work with those indices rather than the types directly.

Reply:
> There will have to be a convention that index 0 of the variant is command queue and index 1 is command list and the user will have to work with those indices rather than the types directly

Are you saying that we would change the public API, so the `std::variant` did not have different types for the two cases?  Or is this just a change at the PI layer?

Even at the PI layer, I don't understand why you would create an `std::variant` when both elements have the same type.  That's not useful.  If the value has a uniform type, you don't need to use `std::variant`.

Reply:
The proposal is to use std::variant with two types that are the same, void* or uint64_t. The index of the variant would tell us what is in it, command queue or command list.

With distinct types in the variant we do this:
// command_queue and command_list are of L0 type.
make_queue({command_queue}, ...); or make_queue({command_list}, ...);
...
auto H=get_native(queue);
if (std::holds_alternative<ze_command_queue_handle_t>(H)) { } else { }

With two types that are the same:
make_queue({std::in_place_index<0>, command_queue}); or make_queue({std::in_place_index<1>, command_list}
...
auto H=get_native(queue);
if (H.index() == 0) {/*command queue*/ } else { /*command list*/}

Reply:
And, to work with the obtained L0 handle from get_native, casts are needed:

if (H.index() == 0) {
  CQ = static_cast<ze_command_queue_handle_t>(std::get<0>(H));
  } else {
  CL = static_cast<ze_command_list_handle_t>(std::get<1>(H));
}

Reply:
This would be a change to the public API that we document in the extension?

That seems less type safe and less easy to use.  What is the advantage to making this change?

Reply:
I don't believe we want this burden (of conversion) exposed to the user, but we should not need L0 specific types in common PI API (they should be abstracted). I admit I don't fully understand why opaque pi_native_handle is not enough.

Reply:
Last I remember, the PI API was pure C.  Is that still the case?  If so, then I think you would not want to use any `std` C++ types in that API.

> I admit I don't fully understand why opaque pi_native_handle is not enough.

I presume you could cast both `ze_command_queue_handle_t` and `ze_command_list_handle_t` to an opaque pointer.  Then you would need some way to tell them apart, like a separate flag or have two different PI APIs.

Reply:
With get_native returning a pointer to two types of objects, command queue or command list, unless it is of type "variant" the caller will still have to do a cast.

So I think we are down to these two choices:
1. use a variant of two opaque types and let the index in the variant distinguish which it is. The user has to check and then do a cast.
2. Go back to my original proposal where the backend type is:
typedef struct _ze_queue_or_cmdlist {
  void *ze_handle;
  bool IsImmCmdList;
} *ze_queue_or_cmdlist_handle_t;

Then a pointer to this struct can be just pass-through to the SYCL headers and library. But here too, the user of get_native has to check the flag and do a cast.

I'm going to update the PR to use #1 and get your feedback on that.

Reply:
can we define the backend type as std::variant<pi_native_handle, pi_native_handle> but wrapped into an object that would have automatic conversions to ze_command_queue_handle_t and ze_command_list_handle_t defined in it?

Reply:
wait, why can't we just do a specialization of the templated `get_native<level_zero, queue>` in level_zero.[hc]pp? like we do for all "make_XXX" ones? this way the L0 types will not be exposed to the rest of the SYCL

Reply:
Can you elaborate? What is pi_native_handle2 defined as?

Reply:
PI will talk in "pi_native_handle" and "bool" language like you already do in: https://github.com/intel/llvm/pull/7196/files#diff-952d2515bc7a8c91d37420f519ff726c0e6ce71f5a81bfd98e5c6fcfcaa3ada0R1204

The "get_native" specialization in level_zero.cpp will covert that to the std::variant that we defined for L0 backend

Reply:
OK, I have made changes that isolate the L0 details to files level_zero.[ch]pp.
The user interface is unchanged.